### PR TITLE
Adjusted code for Xstate V5

### LINKED ISSUE #4: XState v5
XState v5 is in beta. 

I know this should be a PR that is fully working, but I'm on a deadline, and I sort of hacked some changes in to make this work with v5. I am not 100% I have all the proper typings in place. But, maybe this can help creating a `@next` branch to align with the new beta.

```ts

import { atom } from 'jotai/vanilla';
import type { Getter, WritableAtom } from 'jotai/vanilla';
import { interpret } from 'xstate';
import type {
  AnyActorBehavior,
  AnyActorRef,
  AnyInterpreter,
  AnyStateMachine,
  AreAllImplementationsAssumedToBeProvided,
  EventObject,
  InternalMachineImplementations,
  InterpreterFrom,
  InterpreterOptions,
  MachineContext,
  Prop,
  StateConfig,
  StateFrom,
} from 'xstate';

export const RESTART = Symbol();

export interface MachineAtomOptions<
  TContext extends MachineContext,
  TEvent extends EventObject
> {
  /**
   * If provided, will be merged with machine's `context`.
   */
  context?: Partial<TContext>;
  /**
   * The state to rehydrate the machine to. The machine will
   * start at this state instead of its `initialState`.
   */
  state?: StateConfig<TContext, TEvent>;
}

type Options<TMachine extends AnyStateMachine> =
  AreAllImplementationsAssumedToBeProvided<
    TMachine['__TResolvedTypesMeta']
  > extends false
    ? InterpreterOptions<AnyActorBehavior> &
        MachineAtomOptions<TMachine['__TContext'], TMachine['__TEvent']> &
        InternalMachineImplementations<
          TMachine['__TContext'],
          TMachine['__TEvent'],
          TMachine['__TResolvedTypesMeta'],
          true
        >
    : InterpreterOptions<AnyActorBehavior> &
        MachineAtomOptions<TMachine['__TContext'], TMachine['__TEvent']> &
        InternalMachineImplementations<
          TMachine['__TContext'],
          TMachine['__TEvent'],
          TMachine['__TResolvedTypesMeta']
        >;

type MaybeParam<T> = T extends (v: infer V) => unknown ? V : never;

export function atomWithMachine<
  TMachine extends AnyStateMachine,
  TInterpreter = InterpreterFrom<TMachine>
>(
  getMachine: TMachine | ((get: Getter) => TMachine),
  getOptions?: Options<TMachine> | ((get: Getter) => Options<TMachine>)
): WritableAtom<
  StateFrom<TMachine>,
  // [MaybeParam<Prop<TInterpreter, 'send'>> | typeof RESTART],
  [MaybeParam<Prop<TInterpreter, 'send'>> | typeof RESTART],
  void
> {
  const cachedMachineAtom = atom<{
    machine: AnyStateMachine;
    service: AnyInterpreter;
  } | null>(null);
  if (process.env.NODE_ENV !== 'production') {
    cachedMachineAtom.debugPrivate = true;
  }

  const machineAtom = atom(
    (get) => {
      const cachedMachine = get(cachedMachineAtom);
      if (cachedMachine) {
        return cachedMachine;
      }
      let initializing = true;
      const safeGet: typeof get = (...args) => {
        if (initializing) {
          return get(...args);
        }
        throw new Error('get not allowed after initialization');
      };
      const machine = isGetter(getMachine) ? getMachine(safeGet) : getMachine;
      const options = isGetter(getOptions) ? getOptions(safeGet) : getOptions;
      initializing = false;
      const {
        guards,
        actions,
        actors,
        delays,
        context,
        ...interpreterOptions
      } = options || {};

      const machineConfig = {
        ...(guards ? { guards } : { guards: {} }),
        ...(actions ? { actions } : { actions: {} }),
        ...(actors ? { actors } : { actors: {} }),
        ...(delays ? { delays } : { delays: {} }),
      };

      const machineWithConfig = machine.provide({
        ...machineConfig,
        context: {
          ...machine.getContext(),
          ...context,
        },
      } as never); // TODO FIX THIS TYPING

      const service = interpret(machineWithConfig, interpreterOptions);
      return { machine: machineWithConfig, service };
    },
    (get, set) => {
      set(cachedMachineAtom, get(machineAtom));
    }
  );

  machineAtom.onMount = (commit) => {
    commit();
  };

  const cachedMachineStateAtom = atom<StateFrom<TMachine> | null>(null);
  if (process.env.NODE_ENV !== 'production') {
    cachedMachineStateAtom.debugPrivate = true;
  }

  const machineStateAtom = atom(
    (get) =>
      get(cachedMachineStateAtom) ??
      (get(machineAtom).machine.initialState as StateFrom<TMachine>),
    (get, set, registerCleanup: (cleanup: () => void) => void) => {
      const { service } = get(machineAtom);
      service.subscribe((nextState: any) => {
        set(cachedMachineStateAtom, nextState);
      });
      service.start();
      registerCleanup(() => {
        const { service } = get(machineAtom);
        service.stop();
      });
    }
  );

  if (process.env.NODE_ENV !== 'production') {
    machineStateAtom.debugPrivate = true;
  }

  machineStateAtom.onMount = (initialize) => {
    let unsub: (() => void) | undefined | false;

    initialize((cleanup) => {
      if (unsub === false) {
        cleanup();
      } else {
        unsub = cleanup;
      }
    });

    return () => {
      if (unsub) {
        unsub();
      }
      unsub = false;
    };
  };

  const machineStateWithServiceAtom = atom(
    (get) => get(machineStateAtom),
    (
      get,
      set,
      // event: Parameters<AnyInterpreter['send']>[0] | typeof RESTART
      event: Parameters<AnyActorRef['send']>[0] | typeof RESTART
    ) => {
      const { service } = get(machineAtom);
      if (event === RESTART) {
        service.stop();
        set(cachedMachineAtom, null);
        set(machineAtom);
        const { service: newService } = get(machineAtom);
        newService.subscribe((nextState: any) => {
          set(cachedMachineStateAtom, nextState);
        });
        newService.start();
      } else {
        service.send(event);
      }
    }
  );

  return machineStateWithServiceAtom;
}

const isGetter = <T>(v: T | ((get: Getter) => T)): v is (get: Getter) => T =>
  typeof v === 'function';


```

### PR DESCRIPTION
**Pull request by brunocangs**
In this PR I've done the following:

- Updated the XState peer dependency version to ^5
- Added `jotai-effect` to simplify handling state machine subscriptions
- Upgraded the example and tests to work with the new code and follow best practices through the `setup()` function

This PR also closes #5

### CODE REVIEW

**Comment by brunocangs** [comment]
@dai-shi @arjunvegda Let me know if there are any review comments that need adjusting. Thank you

**Comment by dai-shi** [comment]
> * Added `jotai-effect` to simplify handling state machine subscriptions

How hard would it be to implement it here without `jotai-effect`?

**Comment by brunocangs** [comment]
> > * Added `jotai-effect` to simplify handling state machine subscriptions
> 
> How hard would it be to implement it here without `jotai-effect`?

Not hard at all, I used it mostly because I thought the previous `onMount + cleanup` function was a bit hacky and hard to read but I could revert it and it should work just as well

**Comment by brunocangs** [comment]
@dai-shi Update on using `jotai-effect`

Without effect we run into some problems because of `onMount`'s behaviour. I wasn't able to get it to reliably re-trigger the `onMount`, which means that after calling `RESET`, or because of react's strict mode it ends up unsubscribing from the changes in state, causing the whole screen to be unresponsive.

I could keep working but to be it seems a decent tradeoff both in readability, consistency and bundle size considering the gzipped package weights [less than 1kb](https://bundlephobia.com/package/jotai-effect@1.0.0)

gist url: https://gist.github.com/brunocangs/a20305a33c0c48aa7c10dc5b54d3544c

**Comment by dai-shi** [comment]
> Without effect we run into some problems because of `onMount`'s behaviour. I wasn't able to get it to reliably re-trigger the `onMount`,

Haven't thought about it deeply, but I understand there can be a difference. But, it's how it works as of now, right?

That said, I would like to see them as separate PRs.
- This PR: Migrate to XState 5
- Another PR: Fix with jotai-effect

**Comment by brunocangs** [comment]
> That said, I would like to see them as separate PRs.
> 
> * This PR: Migrate to XState 5

Managed to get an implementation that works both in dev and in prod by decoupling the subscription logic into other atoms, so I can reset them in a more predictable way. Would love a review to sanity check it's not causing any weird leaks @dai-shi. 

With this implementation I think the `jotai-effect` isn't necessary

**Comment by brunocangs** [comment]
Fixed tests

**Review by dai-shi — APPROVED** [review]
Not a deep review at all, but the diff looks clean to me.

I wonder if @davidkpiano has a chance to take a look.

**Comment by brunocangs** [comment]
Did some deeper debugging and noticed that the unsubscribe function wasn't getting called on reset. Made some adjustments to fix it.

Tests pass

File: src/atomWithMachine.ts [hunk 1]
```diff
@@ -30,42 +37,27 @@ export interface MachineAtomOptions<TContext, TEvent extends EventObject> {
...
-        InternalMachineOptions<
-          TMachine['__TContext'],
-          TMachine['__TEvent'],
-          TMachine['__TResolvedTypesMeta']
-        >;
+type Options<TMachine extends AnyStateMachine> = Compute<
+  MachineAtomOptions<ContextFrom<TMachine>, EventFrom<TMachine>> &
+    InternalMachineImplementations<
+      ContextFrom<TMachine>,
+      __ResolvedTypesMetaFrom<TMachine>
```
Comment (davidkpiano) [inline src/atomWithMachine.ts:44]: If this `Options` type works, that's great; we're also working on getting rid of typegen types from XState v5 which may simplify this a bit.

**Comment by davidkpiano** [comment]
> Not a deep review at all, but the diff looks clean to me.
> 
> I wonder if @davidkpiano has a chance to take a look.

It generally looks good to me! 💯

File: src/atomWithMachine.ts (hunk 1 above)
Comment (brunocangs) [reply src/atomWithMachine.ts:44]: It seems to be working as far as I can tell too

image

**Comment by dai-shi** [comment]
Our test coverage isn't very good and there might be some regressions. But it's better to move forward than stopping. Let's merge.

**Pull request merged**